### PR TITLE
Allow empty permissions boundary field in AWS role.

### LINKED ISSUE #5546: iam_role: permissions_boundary attribute should accept empty string as absent value
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

```
Terraform v0.11.7
+ provider.aws v1.31.0
+ provider.template v1.0.0
```

### Affected Resource(s)

* aws_iam_role

### Terraform Configuration Files

```hcl
resource "aws_iam_role" "test_role" {
  name = "test_role"
  assume_role_policy = "${data.aws_iam_policy_document.assume_role.json}"
  permissions_boundary = ""
}

data "aws_iam_policy_document" "assume_role" {
  statement {
    actions = ["sts:AssumeRole"]
    effect  = "Allow"

    principals {
      type        = "Service"
      identifiers = ["ec2.amazonaws.com"]
    }
  }
}
```

### Debug Output

Debug log of `terraform plan` : https://gist.github.com/mgarstecki/62cec1b01846e74afbedab1589581c1e

### Expected Behavior

Terraform should generate an execution plan to create the role without a permission boundary.

### Actual Behavior

Terraform fails with the following error :

```
Error: aws_iam_role.test_role: expected length of permissions_boundary to be in the range (20 - 2048), got 
```
### Steps to Reproduce

1. `terraform plan` or `terraform apply`

### Important Factoids

The `permissions_boundary` attribute is optional, and it seems common in TF to treat an empty value as absent for the sake of validation. It looks more like a bug than a feature request to me, but feel free to correct me if this is not how the community would classify this.

Tolerating the empty value is useful in my context : we maintain a module that creates roles, and should work for users with and without a permission boundary constraint. I want to have a `permissions_boundary` input variable to handle both cases:

```hcl
variable "permissions_boundary_arn" {
  default = ""
}

resource "aws_iam_role" "role" {
  name_prefix = "delegated-instance_"

  assume_role_policy = "..."
  permissions_boundary = "${var.permissions_boundary_arn}"
}
```

The current validation rule would oblige me to duplicate the module, one version for users without boundaries, and one version for those that use one.

I'll work on a PR to correct this.

### PR DESCRIPTION
**Pull request by mgarstecki**
Fixes #5546

Changes proposed in this pull request:

* Allow empty value for the `permissions_boundary` attribute of role, and treat it as absent.
* Add a check in role acceptance tests to verify the permissions boundary field in the API on each test step.

Output from acceptance testing:

```
$ make testacc TESTARGS="-run TestAccAWSIAMRole"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run TestAccAWSIAMRole -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSIAMRolePolicy_importBasic
--- PASS: TestAccAWSIAMRolePolicy_importBasic (26.74s)
=== RUN   TestAccAWSIAMRole_importBasic
--- PASS: TestAccAWSIAMRole_importBasic (24.18s)
=== RUN   TestAccAWSIAMRolePolicy_basic
--- PASS: TestAccAWSIAMRolePolicy_basic (42.57s)
=== RUN   TestAccAWSIAMRolePolicy_namePrefix
--- PASS: TestAccAWSIAMRolePolicy_namePrefix (41.43s)
=== RUN   TestAccAWSIAMRolePolicy_generatedName
--- PASS: TestAccAWSIAMRolePolicy_generatedName (41.80s)
=== RUN   TestAccAWSIAMRolePolicy_invalidJSON
--- PASS: TestAccAWSIAMRolePolicy_invalidJSON (1.95s)
=== RUN   TestAccAWSIAMRole_basic
--- PASS: TestAccAWSIAMRole_basic (22.15s)
=== RUN   TestAccAWSIAMRole_basicWithDescription
--- PASS: TestAccAWSIAMRole_basicWithDescription (52.90s)
=== RUN   TestAccAWSIAMRole_namePrefix
--- PASS: TestAccAWSIAMRole_namePrefix (22.32s)
=== RUN   TestAccAWSIAMRole_testNameChange
--- PASS: TestAccAWSIAMRole_testNameChange (53.76s)
=== RUN   TestAccAWSIAMRole_badJSON
--- PASS: TestAccAWSIAMRole_badJSON (2.23s)
=== RUN   TestAccAWSIAMRole_disappears
--- PASS: TestAccAWSIAMRole_disappears (15.59s)
=== RUN   TestAccAWSIAMRole_force_detach_policies
--- PASS: TestAccAWSIAMRole_force_detach_policies (32.12s)
=== RUN   TestAccAWSIAMRole_MaxSessionDuration
--- PASS: TestAccAWSIAMRole_MaxSessionDuration (40.68s)
=== RUN   TestAccAWSIAMRole_PermissionsBoundary
--- PASS: TestAccAWSIAMRole_PermissionsBoundary (89.71s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	510.144s
```

### CODE REVIEW

File: aws/resource_aws_iam_role_test.go [hunk 1]
```diff
@@ -399,6 +412,41 @@ func testAccAddAwsIAMRolePolicy(n string) resource.TestCheckFunc {
...
+			return fmt.Errorf("Not found: %s", n)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("No Role name is set")
+		}
+
+		iamconn := testAccProvider.Meta().(*AWSClient).iamconn
+
+		resp, err := iamconn.GetRole(&iam.GetRoleInput{
```
Comment (bflad) [inline aws/resource_aws_iam_role_test.go]: Instead of making a new API call here, you can just pass in `&role` instead of `resourceName` (which is the same `iam.GetRoleOutput` of `resp`). `testAccCheckAWSRoleExists()` does all the above logic already. 😄

Comment (mgarstecki) [reply aws/resource_aws_iam_role_test.go]: Thank you for the feedback @bflad !
Fixed it, indeed it's much simpler that way :)

**Comment by mgarstecki** [comment]
Implemented @bflad's feedback, thanks !

Here are the new acceptance test results:
```
$ make testacc TESTARGS="-run TestAccAWSIAMRole"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./... -v -run TestAccAWSIAMRole -timeout 120m
?   	github.com/terraform-providers/terraform-provider-aws	[no test files]
=== RUN   TestAccAWSIAMRolePolicy_importBasic
--- PASS: TestAccAWSIAMRolePolicy_importBasic (15.27s)
=== RUN   TestAccAWSIAMRole_importBasic
--- PASS: TestAccAWSIAMRole_importBasic (14.00s)
=== RUN   TestAccAWSIAMRolePolicy_basic
--- PASS: TestAccAWSIAMRolePolicy_basic (24.76s)
=== RUN   TestAccAWSIAMRolePolicy_namePrefix
--- PASS: TestAccAWSIAMRolePolicy_namePrefix (23.82s)
=== RUN   TestAccAWSIAMRolePolicy_generatedName
--- PASS: TestAccAWSIAMRolePolicy_generatedName (23.64s)
=== RUN   TestAccAWSIAMRolePolicy_invalidJSON
--- PASS: TestAccAWSIAMRolePolicy_invalidJSON (1.21s)
=== RUN   TestAccAWSIAMRole_basic
--- PASS: TestAccAWSIAMRole_basic (12.59s)
=== RUN   TestAccAWSIAMRole_basicWithDescription
--- PASS: TestAccAWSIAMRole_basicWithDescription (30.20s)
=== RUN   TestAccAWSIAMRole_namePrefix
--- PASS: TestAccAWSIAMRole_namePrefix (12.19s)
=== RUN   TestAccAWSIAMRole_testNameChange
--- PASS: TestAccAWSIAMRole_testNameChange (29.82s)
=== RUN   TestAccAWSIAMRole_badJSON
--- PASS: TestAccAWSIAMRole_badJSON (1.25s)
=== RUN   TestAccAWSIAMRole_disappears
--- PASS: TestAccAWSIAMRole_disappears (8.74s)
=== RUN   TestAccAWSIAMRole_force_detach_policies
--- PASS: TestAccAWSIAMRole_force_detach_policies (17.39s)
=== RUN   TestAccAWSIAMRole_MaxSessionDuration
--- PASS: TestAccAWSIAMRole_MaxSessionDuration (23.05s)
=== RUN   TestAccAWSIAMRole_PermissionsBoundary
--- PASS: TestAccAWSIAMRole_PermissionsBoundary (49.63s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	287.576s
```

**Comment by bflad** [comment]
This has been released in [version 1.35.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1350-september-06-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
